java/ql/lib/semmle/code/java/dataflow/internal/DataFlowImpl.qll

```diff
@@ -2011,617 +1788,430 @@ private module Stage3 {
-    clearContent(node, ap.getHead().getContent(), config)
-  }
+  module NoLocalCallContext {
+    class LocalCc = Unit;
 
-  pragma[nomagic]
-  private predicate expectsContentCand(NodeEx node, Ap ap, Configuration config) {
-    exists(Content c |
-      PrevStage::revFlow(node, pragma[only_bind_into](config)) and
-      PrevStage::readStepCand(_, c, _, pragma[only_bind_into](config)) and
-      expectsContentEx(node, c) and
-      c = ap.getHead().getContent()
-    )
+    bindingset[node, cc]
+    LocalCc getLocalCc(NodeEx node, Cc cc) { any() }
+
+    bindingset[call, c, outercc]
+    CcCall getCallContextCall(DataFlowCall call, DataFlowCallable c, Cc outercc) {
+      checkCallContextCall(outercc, call, c) and
+      if recordDataFlowCallSiteDispatch(call, c)
+      then result = TSpecificCall(call)
+      else result = TSomeCall()
+    }
   }
 
-  pragma[nomagic]
-  private predicate castingNodeEx(NodeEx node) { node.asNode() instanceof CastingNode }
+  module LocalCallContext {
+    class LocalCc = LocalCallContext;
 
-  bindingset[node, state, ap, config]
-  private predicate filter(NodeEx node, FlowState state, Ap ap, Configuration config) {
-    exists(state) and
-    exists(config) and
-    not clear(node, ap, config) and
-    (if castingNodeEx(node) then compatibleTypes(node.getDataFlowType(), ap.getType()) else any()) and
-    (
-      notExpectsContent(node)
-      or
-      expectsContentCand(node, ap, config)
-    )
-  }
+    bindingset[node, cc]
+    LocalCc getLocalCc(NodeEx node, Cc cc) {
+      result =
+        getLocalCallContext(pragma[only_bind_into](pragma[only_bind_out](cc)),
+          node.getEnclosingCallable())
+    }
 
-  bindingset[ap, contentType]
-  private predicate typecheckStore(Ap ap, DataFlowType contentType) {
-    // We need to typecheck stores here, since reverse flow through a getter
-    // might have a different type here compared to inside the getter.
-    compatibleTypes(ap.getType(), contentType)
+    bindingset[call, c, outercc]
+    CcCall getCallContextCall(DataFlowCall call, DataFlowCallable c, Cc outercc) {
+      checkCallContextCall(outercc, call, c) and
+      if recordDataFlowCallSite(call, c) then result = TSpecificCall(call) else result = TSomeCall()
+    }
   }
 
-  /* Begin: Stage 3 logic. */
-  bindingset[node, state, config]
-  private predicate flowCand(NodeEx node, FlowState state, ApApprox apa, Configuration config) {
-    PrevStage::revFlow(node, state, _, _, apa, config)
+  bindingset[call, c, innercc]
+  CcNoCall getCallContextReturn(DataFlowCallable c, DataFlowCall call, Cc innercc) {
+    checkCallContextReturn(innercc, c, call) and
+    if reducedViableImplInReturn(c, call) then result = TReturn(c, call) else result = ccNone()
   }
+}
 
-  bindingset[result, apa]
-  private ApApprox unbindApa(ApApprox apa) {
-    pragma[only_bind_out](apa) = pragma[only_bind_out](result)
-  }
+private module Stage2Param implements MkStage<Stage1>::StageParam {
```

Comment:
We have two calls to `MkStage<Stage1>` (and same for other stages), here and in the module `Stage2`. Perhaps that is OK, as parameterized modules are not generative?

Reply:
That is indeed perfectly fine.